DemocracyClub/uk-geo-utils
# Feature/onspd header checks and db flag

First commit sorts out the ONSPD header issue and does some nice checks. When it doesn't match we will get output that looks something like: 
```
CommandError: 
Problem with the fields in /home/will/Downloads/ONSPD_MAY_2024_UK/ONSPD_MAY_2024_UK.csv
  File header: ['bua11', 'buasd11', 'calncv', 'casward', 'ced', 'ctry', 'dointr', 'doterm', 'eer', 'icb', 'imd', 'itl', 'lat', 'lep1', 'lep2', 'long', 'lsoa01', 'lsoa11', 'lsoa21', 'msoa01', 'msoa11', 'msoa21', 'nhser', 'npark', 'oa01', 'oa11', 'oa21', 'oac01', 'oac11', 'oscty', 'oseast1m', 'osgrdind', 'oshlthau', 'oslaua', 'osnrth1m', 'osward', 'parish', 'pcd', 'pcd2', 'pcds', 'pcon', 'pct', 'pfa', 'rgn', 'ru11ind', 'sicbl', 'statsward', 'streg', 'teclec', 'ttwa', 'ur01ind', 'usertype', 'wz11']
  Expected header: ['bua22', 'calncv', 'casward', 'ced', 'ctry', 'dointr', 'doterm', 'eer', 'icb', 'imd', 'itl', 'lat', 'lep1', 'lep2', 'long', 'lsoa01', 'lsoa11', 'lsoa21', 'msoa01', 'msoa11', 'msoa21', 'nhser', 'npark', 'oa01', 'oa11', 'oa21', 'oac01', 'oac11', 'oscty', 'oseast1m', 'osgrdind', 'oshlthau', 'oslaua', 'osnrth1m', 'osward', 'parish', 'pcd', 'pcd2', 'pcds', 'pcon', 'pct', 'pfa', 'rgn', 'ru11ind', 'sicbl', 'statsward', 'streg', 'teclec', 'ttwa', 'ur01ind', 'usertype', 'wz11']
  Fields missing from file:
  - bua22
  Unexpected fields found in file:
  + bua11
  + buasd11
This probably means ONSPD has changed they're csv format and we need to update our import command.
```

The second commit adds a `--database` flag to the `BaseImporter` meaning that classes that inherit can target specific databases. I've added some tests for this, because it's easy not to use the `BaseImporter`'s cursor (i.e. `self.cursor`). 

To test locally you can:
- got to (eg) everyelection and `pip install -e /path/to/uk-geo-utils/`
- Duplcate your existing db `createdb -T every_election every_election_20241111_test ` 
- add this to your local.py `DATABASES` setting:
  ```
      "other": {
        "ENGINE": "django.contrib.gis.db.backends.postgis",
        "NAME": "every_election_20241111_test",
        "USER": "every_election",
        "PASSWORD": "",
        "HOST": "",
        "PORT": "",
    },
    ```
- This should fail: `python manage.py import_onspd --data-path /home/will/Downloads/ONSPD_AUG_2024/Data/`
- migrate `python manage.py migrate`. Now the above command should succeed
- This will fail: `python manage.py import_onspd --data-path /home/will/Downloads/ONSPD_AUG_2024/Data/ --database other`
- `python manage.py migrate --database other`  Now it should succeed.

Between the above commands you can check where the data is being imported with   `select count(*) from uk_geo_utils_onspd;` in the relevant dbs.
